### PR TITLE
Tests: ngx_http_json_module tests.

### PR DESCRIPTION
**Pull request by roman-f5**
### Proposed changes

Added json_parser.t covering json_set value extraction (scalars, nested
objects and arrays, compound containers, literals, number types, escapes,
the '$' root selector, source caching, last-match-wins on duplicate keys,
quoted key segments, and binary-safe / Unicode / surrogate-pair keys), and
json_parser_invalid.t covering rejection of malformed json_set paths at
configuration load.

### Checklist

Before creating a PR, run through this checklist and mark each as complete:

- [x] I have read the [contributing guidelines](/CONTRIBUTING.md).
- [x] I have signed the [F5 Contributor License Agreement (CLA)](https://github.com/f5/f5-cla/blob/main/docs/f5_cla.md).
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works.
- [x] If applicable, I have checked that any relevant tests pass after adding my changes.
- [x] I have updated any relevant documentation ([`README.md`](/README.md) and/or [`CHANGELOG.md`](/CHANGELOG.md)).

### CODE REVIEW

File: json_parser.t [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+
+        location /surrogate {
+            return 200 "sur=[$jp_qk_sur]\n";
+        }
+    }
+}
+
+EOF
+
+$t->try_run('no json_set')->plan(45);
```
Comment (dekobon) [inline json_parser.t]: This is good coverage for `json_set` path/extraction behavior, but it is not yet a parser conformance test.

The current file has 29 runtime extraction cases and only one malformed JSON input (`{bad`). That means a large class of parser regressions could still land unnoticed: malformed numbers, trailing garbage, invalid UTF-8, lone or mispaired surrogates, BOM handling, unusual whitespace, deep nesting, and other edge cases that are often security-relevant because different parsers disagree on them.

JSONTestSuite is a public conformance corpus used by many JSON parser implementations. It includes:
- 95 valid cases (`y_*`)
- 188 invalid cases (`n_*`)
- 35 implementation-defined edge cases (`i_*`)
- 22 transform cases

Its value is that it exercises exactly the kinds of parser boundary cases that hand-written happy-path tests usually miss.

I think this PR needs a second layer of tests:
1. keep `json_parser.t` for module behavior and path extraction
2. add a direct parser conformance test around `ngx_json_validate()` that runs a representative subset, or ideally the full JSONTestSuite corpus

A direct parser test is probably the better long-term place for this because some important JSONTestSuite inputs use raw bytes (invalid UTF-8, BOMs, embedded NULs, malformed escapes) that are awkward or impossible to express reliably through `$arg_json` and HTTP query-string based tests.

File: json_set.t [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+    json_set $jp_score0 $arg_json results.scores[0];
+    json_set $jp_score1 $arg_json results.scores[1];
+    json_set $jp_scores $arg_json results.scores;
+
+    json_set $jp_bt $arg_json bt;
+    json_set $jp_bf $arg_json bf;
+    json_set $jp_nv $arg_json nv;
+
+    json_set $jp_deep $arg_json a.b.c.d.e;
```
Comment (dekobon) [inline json_set.t:64]: There is no `json_max_depth` coverage in this file yet. Since the directive directly controls the parser stack bound, it would be useful to add at least:
- `json_max_depth 1` with a one-level success case
- `json_max_depth 1` with a two-level rejection case
- `json_max_depth 0` config rejection

Once the upper-bound fix lands, a huge-value config rejection should be covered too. These are exactly the boundary tests that would catch regressions in depth enforcement and the stack-allocation path.

File: json_set.t [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+    json_set $jp_ov_obj $arg_json a.b;
+    json_set $jp_ov_c $arg_json a.b.c;
+
+    json_set $jp_ar0 $arg_json arr[0];
+    json_set $jp_ar2 $arg_json arr[2];
+
+    json_set $jp_loc_name $arg_json name;
+
+    json_set $jp_cache_a $arg_json a;
```
Comment (dekobon) [inline json_set.t:103]: Could we add regression coverage for destination-variable interference here? All current destinations are fresh `$jp_*` names, so this file would not catch either sibling-slot clobbering or collisions with built-in prefix variables.

Two small cases would help a lot:
- `set $x ...; json_set $x ...; json_set $y ...; return "$x$y";` to ensure evaluating `$y` does not rewrite `$x`
- `json_set $http_authorization ...;` with and without an `Authorization` header to make the intended collision behavior explicit

Those are security-relevant because they can change auth-related values based only on evaluation order or destination naming.

File: json_parser.t [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+	open my $fh, '>', $file or die "can't open $file: $!";
+	print $fh $conf;
+	close $fh;
+
+	# use a dedicated log file so these nginx -t runs never write into the
+	# error.log that the Test::Nginx teardown inspects for alerts
+	my $bin = $ENV{TEST_NGINX_BINARY} || 'nginx';
+	my $out = `$bin -t -p $testdir/ -c invalid.conf -e invalid_error.log 2>&1`;
+
+	# ignore the nginx -t exit status: an otherwise-valid config may still
```
Comment (dekobon) [inline json_parser.t]: `check_conf()` ignores the `nginx -t` exit status and decides only from regex output. That is fine for rejection checks, but it makes accepted-path checks like `foo["x"].bar` capable of false-passing if `nginx -t` fails for some unrelated reason. Could we split this into `check_conf_rejected()` and `check_conf_ok()` helpers, with the acceptance helper requiring exit status 0?

File: json_set.t [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+like(http_get('/quoted?json={"foo":{"x[0]":"brkkey"},'
+	. '"a]b":"rbrkey","":"emptykey","data":{"a.b":{"c":"deep"}}}'),
+	qr/brk=\[brkkey\] rbr=\[rbrkey\] empty=\[emptykey\] mixed=\[deep\]/,
+	'quoted key segments');
+
+like(http_get('/quoted?json={"a.b":"x","foo":{},"a]b":"y","data":{}}'),
+	qr/empty=\[\]/, 'quoted empty key absent');
+
+like(http_post('/allq', '{"\\"asd\\u8898  smth\\"":"hit"}'),
```
Comment (dekobon) [inline json_set.t:517]: The `$request_body` coverage here only uses small in-memory bodies. Could we add a spill regression as well? With a low `client_body_buffer_size`, the same logical JSON body can spill to a temp file and make `$request_body`-backed `json_set` values disappear. That is security-relevant for configs that use an extracted field for allow/deny decisions.

A useful matrix would be:
- small `{"decision":"deny"}` body -> extracted value present
- same body padded past the spill threshold -> policy result must not silently change
- same padded body with a larger body buffer -> extracted value present again

That would catch fail-open behavior caused only by body size / storage mode.

File: json_parser.t (hunk 4 above)
Comment (dekobon) [reply json_parser.t]: Correction to my earlier suggestion: requiring exit status 0 is not robust on sanitizer builds, because a valid `nginx -t` can print `syntax is ok` and still exit non-zero after LeakSanitizer output.

The core issue is still that accepted-path checks currently pass on any output that lacks `invalid json_set path`; a better accept helper would require `syntax is ok` and `diag($out)` on mismatch.

File: json_parser.t [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+		. "}\n";
+
+	my $file = "$testdir/invalid.conf";
+	open my $fh, '>', $file or die "can't open $file: $!";
+	print $fh $conf;
+	close $fh;
+
+	# use a dedicated log file so these nginx -t runs never write into the
+	# error.log that the Test::Nginx teardown inspects for alerts
+	my $bin = $ENV{TEST_NGINX_BINARY} || 'nginx';
```
Comment (dekobon) [inline json_parser.t]: `check_conf()` is not using the same nginx binary as the rest of the test in the default workflow. `Test::Nginx` falls back to `../nginx/objs/nginx`, but this falls back to bare `nginx` from PATH. With `TEST_NGINX_BINARY` unset I reproduced the HTTP cases running against the local build while 15 config checks ran against `/usr/sbin/nginx` and failed with `unknown directive "json_set"`; the acceptance check still passed vacuously.

Could this use `$Test::Nginx::NGINX` instead? It also avoids the `||` vs `defined` mismatch for an empty-but-set env var.

File: json_set.t [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+like(http_get('/escaped?json={"escaped":"a\\nb\\tc"}'), qr/esc=\[a\nb\tc\]/,
+	'string with escapes');
+
+like(http_get('/empty?json={"eobj":{},"earr":[]}'),
+	qr/eobj=\[\{\}\] earr=\[\[\]\]/, 'empty object and array');
+
+like(http_get('/toparray?json=[10,20,30]'), qr/t0=\[10\] t1=\[20\]/,
+	'top-level array');
+
+like(http_get('/invalid?json={bad'), qr/name=\[\]/, 'invalid json');
```
Comment (dekobon) [inline json_set.t:392]: The only invalid JSON case fails before any destination is stored, so it does not exercise the second reset after a later parse failure.

A regression that drops the post-failure reset would still pass this file while leaking `name=[John]` from inputs like `{"name":"John",` or `{"name":"John","x":"\uDEAD"}`.

Could we add one truncated-after-match case and one invalid-sibling-string case?

File: json_parser.t [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+	qr/eobj=\[\{\}\] earr=\[\[\]\]/, 'empty object and array');
+
+like(http_get('/toparray?json=[10,20,30]'), qr/t0=\[10\] t1=\[20\]/,
+	'top-level array');
+
+like(http_get('/invalid?json={bad'), qr/name=\[\]/, 'invalid json');
+
+like(http_get('/empty_src'), qr/name=\[\]/, 'empty source');
+
+like(http_get('/loc?json={"name":"locval"}'), qr/loc_name=\[locval\]/,
```
Comment (dekobon) [inline json_parser.t]: This does not test a location-level directive: `$jp_loc_name` is declared in `http {}` like the other vars, and `json_set` itself is `NGX_HTTP_MAIN_CONF` only. `/loc` is therefore just another `name` extraction case.

Could this become a deliberate same-path multi-destination test plus a separate `check_conf()` case asserting that `json_set` is rejected inside `server {}` / `location {}`?

File: json_parser.t [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+	'location level directive');
+
+like(http_get('/cache?json={"a":"one","b":"two"}'), qr/a=\[one\] b=\[two\]/,
+	'cache shared source');
+
+like(http_get('/skipidx?json=[{"ignored":1},{"wanted":"yes"}]'),
```
Comment (dekobon) [inline json_parser.t]: This input does not actually skip element 0: `[0]` is already configured on the same `$arg_json` tree via `$jp_top0`, so the first element is matched and descended.

A skip-on-open regression would leave this green. Could this use a dedicated source/path with only index 1 configured, e.g. `arrx[1].wanted`, so element 0 is genuinely unconfigured?

File: json_set.t [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+	'deeply nested path');
+
+like(http_get('/matrix?json={"matrix":[[1,2],[3,4]]}'),
+	qr/m00=\[1\] m01=\[2\] m10=\[3\] m1=\[\[3,4\]\]/, 'array of arrays');
+
+like(http_get('/numbers?json={"num_int":42,"num_neg":-7,"num_frac":3.14,'
+	. '"num_exp":1e10}'), qr/int=\[42\] neg=\[-7\] frac=\[3.14\] exp=\[1e10\]/,
+	'number types');
+
+like(http_get('/escaped?json={"escaped":"a\\nb\\tc"}'), qr/esc=\[a\nb\tc\]/,
```
Comment (dekobon) [inline json_set.t:365]: All current `\u` assertions are key-side, where both the config path and the JSON key are decoded by the same unescaper before comparison.

That means a systematically wrong UTF-8 encoding could still pass. Could we add a value-side assertion such as `{"e":"\uD83D\uDE00"}` -> decoded emoji, and perhaps `\u8898`, to pin the actual output bytes?

File: json_set.t [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+is(check_path('foo.'), 1, 'trailing dot rejected');
+is(check_path('foo..bar'), 1, 'double dot rejected');
+is(check_path('$.foo'), 1, 'root selector followed by segment rejected');
+
+is(check_path('a[]'), 1, 'empty index rejected');
+is(check_path('a[xyz]'), 1, 'non-digit index rejected');
+is(check_path('a[1a]'), 1, 'index with trailing letter rejected');
+is(check_path('a[12'), 1, 'unterminated index rejected');
+
+is(check_path('a["abc'), 1, 'unterminated quoted key rejected');
```
Comment (dekobon) [inline json_set.t:551]: This rejection block only covers structural path syntax. It misses the quoted-key unescape validator (`a["x\q"]`, `a["\u12"]`, `o["x\uD83D"]`), the empty-path branch (which reports `empty json_set path`, so `check_path()` cannot match it), and non-`$` name/source arguments (`invalid variable name`).

Those are separate config-time rejection paths worth pinning here.

File: json_parser.t (hunk 1 above)
Comment (dekobon) [inline json_parser.t]: `try_run('no json_set')` turns any startup error in this full config into `skip_all`, not just an absent module.

Because this config contains the hardest quoted-key / surrogate / NUL cases, a regression that makes one of them fail config load would make CI green by skip. Could this first probe module presence with a minimal one-directive config, then call `run()` on the full config so real config regressions fail?

File: json_set.t [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+
+    json_set $jp_loc_name $arg_json name;
+
+    json_set $jp_cache_a $arg_json a;
+    json_set $jp_cache_b $arg_json b;
+
+    json_set $jp_dup $arg_json dup;
+    json_set $jp_dup_nested $arg_json obj.k;
+
+    json_set $jp_nul_x $request_body '["a\\u0000x"]';
```
Comment (dekobon) [inline json_set.t:114]: There are two parser sources in the config (`$arg_json` and `$request_body`), but no request evaluates vars from both sources together.

That leaves per-source reset/isolation untested. A mixed location that returns one `$arg_json`-derived value alongside the `$request_body`-derived values would catch a regression that resets every source's destinations instead of only the current source.

File: json_parser.t [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+	'literals true false null');
+
+like(http_get('/deep?json={"a":{"b":{"c":{"d":{"e":42}}}}}'), qr/deep=\[42\]/,
+	'deeply nested path');
+
+like(http_get('/matrix?json={"matrix":[[1,2],[3,4]]}'),
+	qr/m00=\[1\] m01=\[2\] m10=\[3\] m1=\[\[3,4\]\]/, 'array of arrays');
+
+like(http_get('/numbers?json={"num_int":42,"num_neg":-7,"num_frac":3.14,'
+	. '"num_exp":1e10}'), qr/int=\[42\] neg=\[-7\] frac=\[3.14\] exp=\[1e10\]/,
```
Comment (dekobon) [inline json_parser.t]: Two assertions here are weaker than intended:

- `qr/frac=\[3.14\]/` leaves `.` unescaped, so `frac=[3X14]` would still match.
- The quoted-empty-key case later only checks `empty=[]`, which would also pass if the whole parse failed.

Please use `3\.14` here and anchor the empty-key case with a present sibling, e.g. `rbr=[y] empty=[]`.

**Comment by dekobon** [comment]
The PR description no longer matches the diff: it says there is a separate `json_parser_invalid.t` file, but this PR only adds `json_parser.t`, and it lists `$` root selector extraction even though the test correctly asserts that `$` is rejected.

Could the description be updated to match the current coverage?

File: json_parser.t (hunk 1 above)
Comment (roman-f5) [reply json_parser.t]: you can find the test results below

[parsing.html](https://github.com/user-attachments/files/31564790/parsing.html)
[style.css](https://github.com/user-attachments/files/31564792/style.css)

File: json_set.t (hunk 2 above)
Comment (roman-f5) [reply json_set.t:64]: tests added

File: json_set.t (hunk 3 above)
Comment (roman-f5) [reply json_set.t:103]: tests added

File: json_parser.t (hunk 4 above)
Comment (roman-f5) [reply json_parser.t]: implemented

File: json_parser.t (hunk 6 above)
Comment (roman-f5) [reply json_parser.t]: fixed

File: json_parser.t (hunk 8 above)
Comment (roman-f5) [reply json_parser.t]: fixed

File: json_set.t (hunk 7 above)
Comment (roman-f5) [reply json_set.t:392]: tests added

File: json_parser.t (hunk 9 above)
Comment (roman-f5) [reply json_parser.t]: fixed

File: json_set.t (hunk 10 above)
Comment (roman-f5) [reply json_set.t:365]: fixed

File: json_set.t (hunk 11 above)
Comment (roman-f5) [reply json_set.t:551]: tests added

File: json_parser.t (hunk 1 above)
Comment (roman-f5) [reply json_parser.t]: fixed

File: json_parser.t (hunk 13 above)
Comment (roman-f5) [reply json_parser.t]: fixed

File: json_set.t (hunk 12 above)
Comment (roman-f5) [reply json_set.t:114]: fixed

File: json_set.t (hunk 5 above)
Comment (roman-f5) [reply json_set.t:517]: Good catch.

I tried adding this and it currently fails for a real reason: when the body
spills to a temp file, $request_body-backed json_set goes empty. Since this PR is only
adding tests, I did not leave a permanently failing spill regression in place.

Comment (roman-f5) [reply json_set.t:517]: let's document behavior.

File: json_parser.t [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+
+        location / {
+            return 200 "v=[$v]\n";
+        }
+    }
+}
+
+EOF
+
+$probe->try_run('no json_set');
```
Comment (pluknet) [inline json_parser.t]: you don't need try_run(), it can be replaced with a more precise Test::Nginx::has() test:
```
my $t = Test::Nginx->new()->has(qw/http json rewrite proxy/)->plan(65);
```

As "json" is quite a common name, you may want to amend Test::Nginx::has_module() with appropriate `%regex` entry.

This will allow to simplify the prereq check, removing the whole "probe" block.

File: json_set.t [hunk 15]
```diff
@@ -0,0 +1,648 @@
...
+# checks use standalone "nginx -t" helpers for explicit accept and reject
+# cases.
+
+###############################################################################
+
+use warnings;
+use strict;
+
+use Test::More;
+BEGIN { use FindBin; chdir($FindBin::Bin); }
```
Comment (pluknet) [inline json_set.t:14]: empty line here please

File: json_parser.t [hunk 16]
```diff
@@ -0,0 +1,648 @@
...
+###############################################################################
+
+use warnings;
+use strict;
+
+use Test::More;
+BEGIN { use FindBin; chdir($FindBin::Bin); }
+
+use lib 'lib';
+use Test::Nginx qw/ http http_get /;
```
Comment (pluknet) [inline json_parser.t]: since both methods are exported by default, you can omit listing them explicitly (this is how it is used in other test files)

File: json_parser.t [hunk 17]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+EOF
+
+$t->run()->plan(65);
+
+# capture what check_path()/check_conf() need as plain values so the subs do
+# not close over $t (a closure would keep $t alive past the plan check and
+# reorder the teardown tests)
+my $testdir = $t->{_testdir};
```
Comment (pluknet) [inline json_parser.t]: please you the `$t->testdir()` accessor

File: json_parser.t [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+    }
+}
+
+EOF
+
+$t->run()->plan(65);
+
+# capture what check_path()/check_conf() need as plain values so the subs do
+# not close over $t (a closure would keep $t alive past the plan check and
+# reorder the teardown tests)
```
Comment (pluknet) [inline json_parser.t]: That's kind of "inner knowledge" you could omit these comments, up to you though

**Review by pluknet — COMMENTED** [review]
please rename json_parset.t to json_set.t - to follow the naming convention of using the (main) directive name as exported by the tested module

I didn't look deeply into the essence of the tests, especially into the syntax-related.
If you believe they are truly needed and they pass on all tested platforms, I'm ok to keep them.

**Comment by route443** [comment]
It's time to take care of these tests a bit! Roman -> Ivan

File: json_parser.t (hunk 14 above)
Comment (route443) [reply json_parser.t]: Thanks for the review @pluknet!
I've removed the entire `$probe` block and the `try_run()` call + added the json entry for `has()`

File: json_set.t (hunk 15 above)
Comment (route443) [reply json_set.t:14]: Thanks, fixed!

File: json_parser.t (hunk 16 above)
Comment (route443) [reply json_parser.t]: Fixed

File: json_parser.t (hunk 17 above)
Comment (route443) [reply json_parser.t]: Fixed

File: json_parser.t (hunk 18 above)
Comment (route443) [reply json_parser.t]: Adjusted all comments to align +/- with the general nginx-tests approach

**Review by route443 — APPROVED** [review]
All of pluknet's comments are addressed.
Plus the coverage gaps and two defects in the config-check helper:

- tests added to cover whitespace characters (`ngx_json_ws()` branch)
- certain num vals ​​and exponent sign processing
- top-level scalars
- expanded malformed docs tests
- escape decoding
- empty string val
- some config time tests

Plus some issues with config checks:
- `run_check_conf()` writes the config using standard `open`/`print` calls, bypassing `write_file_expand()` => `listen 127.0.0.1:8080` directive within it is not replaced with the allocated port => `prove -j` = fail
- `server {}` block is unnecessary and removed

Now we have 102 tests. Coverage:

file | strings | branches
-- | -- | --
ngx_http_json_module.c | 92.33% (391) | 82.05% (234)
ngx_json_parse.c | 89.46% (408) | 79.72% (355)
ngx_json_unescape.c | 96.77% (124) | 85.94% (64)

Delta relative to the previous coverage: +1.8 / +11.8 / +15.3 by line, +1.7 / +17.8 / +17.2 by branch.

I made a few adjustments to the sandbox CI so that the json module builds by default and there are no checks in `has()`.  All tests are passing successfully.
